File: src/scenarios/dynamicDaisyChain/tunnelGeometry.ts
import { createRng } from "./config";
import type { ScenarioVariantId, TunnelSegment, TunnelGeometry } from "./types";

function jitter(rng: () => number, base: number, spread: number): number {
  return base + (rng() - 0.5) * 2 * spread;
}

export function buildTunnelGeometry(variant: ScenarioVariantId["id"], seed: number): TunnelGeometry {
  const rng = createRng(seed ^ 0x9e3779b9);
  const lengthBase =
    variant === "deep" ? 260 : variant === "narrow" || variant === "collapsed" ? 140 : 195;
  const lengthM = jitter(rng, lengthBase, 12);
  const widthM = variant === "narrow" ? 2.1 + rng() * 0.4 : 3.2 + rng() * 0.6;

  const segments: TunnelSegment[] = [];
  let cursor = 0;
  const chunk = lengthM / (5 + Math.floor(rng() * 3));
  const labels = ["ingress", "choke_a", "mid_mine", "choke_b", "deep_shaft"];
  for (let i = 0; i < 5 && cursor < lengthM - 1; i++) {
    const end = Math.min(lengthM, cursor + chunk * (0.85 + rng() * 0.35));
    segments.push({
      id: labels[i] ?? `seg_${i}`,
      startS: cursor,
      endS: end,
      label: labels[i] ?? `seg_${i}`,
      attenuationMul: 1 + rng() * (variant === "collapsed" ? 0.45 : 0.22),
      blocked: variant === "collapsed" && i === 3 && rng() > 0.35,
    });
    cursor = end;
  }
  if (segments.length && segments[segments.length - 1].endS < lengthM) {
    const last = segments[segments.length - 1];
    segments.push({
      id: "terminal",
      startS: last.endS,
      endS: lengthM,
      label: "terminal",
      attenuationMul: 1.05 + rng() * 0.15,
      blocked: false,
    });
  }

  const collapsePoints: number[] = [];
  for (let i = 0; i < 3; i++) {
    collapsePoints.push(jitter(rng, lengthM * (0.25 + i * 0.22), 8));
  }

  const relayAnchorZones: TunnelGeometry["relayAnchorZones"] = [];
  for (let z = 0; z < 4; z++) {
    const center = lengthM * (0.18 + z * 0.2) + (rng() - 0.5) * 10;
    relayAnchorZones.push({
      id: `anchor_${z}`,
      startS: Math.max(0, center - 4 - rng() * 3),
      endS: Math.min(lengthM, center + 4 + rng() * 3),
    });
  }

  const signalShadowZones: TunnelGeometry["signalShadowZones"] = [];
  for (let s = 0; s < 2; s++) {
    const c = lengthM * (0.35 + s * 0.28);
    signalShadowZones.push({
      startS: c - 9,
      endS: c + 9,
      lossAdd: 0.08 + rng() * (variant === "noisy" ? 0.18 : 0.1),
    });
  }

  const frontierPoints: number[] = [];
  for (let f = 0; f < 5; f++) {
    frontierPoints.push(jitter(rng, lengthM * (0.12 + f * 0.16), 6));
  }

  const targetZones =
    variant === "target_rich"
      ? [
          { id: "victim_a", startS: lengthM * 0.55, endS: lengthM * 0.62 },
          { id: "victim_b", startS: lengthM * 0.72, endS: lengthM * 0.8 },
        ]
      : [{ id: "victim_zone", startS: lengthM * 0.68, endS: lengthM * 0.76 }];

  const checkpoints = [lengthM * 0.12, lengthM * 0.35, lengthM * 0.55];

  return {
    lengthM,
    widthM,
    entranceS: 0,
    segments,
    collapsePoints,
    relayAnchorZones,
    signalShadowZones,
    frontierPoints,
    targetZones,
    returnCheckpoints: checkpoints,
  };
}

File: controllers/src/tunnel_geometry.py
"""Procedural tunnel geometry (200 m axis) aligned with Track 2 Dynamic Daisy Chain."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List
import random


@dataclass
class TunnelSegment:
    id: str
    start_s: float
    end_s: float
    label: str
    attenuation_mul: float
    blocked: bool = False


@dataclass
class TunnelGeometry:
    length_m: float = 200.0
    width_m: float = 10.0
    height_m: float = 5.0
    entrance_s: float = 0.0
    segments: List[TunnelSegment] = field(default_factory=list)
    collapse_points: List[float] = field(default_factory=list)
    relay_anchor_zones: List[tuple[float, float, str]] = field(default_factory=list)
    signal_shadow_zones: List[tuple[float, float, float]] = field(default_factory=list)
    target_zones: List[tuple[float, float, str]] = field(default_factory=list)


def _jitter(rng: random.Random, base: float, spread: float) -> float:
    return base + (rng.random() - 0.5) * 2 * spread


def build_tunnel_geometry(seed: int = 42) -> TunnelGeometry:
    rng = random.Random(seed ^ 0x9E3779B9)
    length_m = _jitter(rng, 200.0, 8.0)
    width_m = 9.0 + rng.random() * 1.2
    segments: List[TunnelSegment] = []
    chunk = length_m / 5.5
    cursor = 0.0
    labels = ["ingress", "dust_mid", "choke", "deep_shaft", "terminal"]
    for i, lab in enumerate(labels):
        if cursor >= length_m - 0.5:
            break
        span = chunk * (0.85 + rng.random() * 0.35)
        end = min(length_m, cursor + span)
        segments.append(
            TunnelSegment(
                id=f"seg_{i}",
                start_s=cursor,
                end_s=end,
                label=lab,
                attenuation_mul=1.0 + rng.random() * 0.35 + (0.25 if i >= 2 else 0),
                blocked=i == 3 and rng.random() > 0.65,
            )
        )
        cursor = end
    if segments and segments[-1].end_s < length_m:
        last = segments[-1]
        segments.append(
            TunnelSegment(
                id="terminal_pad",
                start_s=last.end_s,
                end_s=length_m,
                label="terminal",
                attenuation_mul=1.08 + rng.random() * 0.12,
                blocked=False,
            )
        )
    collapse_points = [length_m * (0.22 + i * 0.21) + (rng.random() - 0.5) * 6 for i in range(4)]
    relay_anchor_zones = []
    for z in range(4):
        c = length_m * (0.15 + z * 0.22) + (rng.random() - 0.5) * 8
        relay_anchor_zones.append((max(0.0, c - 5), min(length_m, c + 5), f"anchor_{z}"))
    signal_shadow_zones = []
    for s in range(3):
        c = length_m * (0.28 + s * 0.24)
        signal_shadow_zones.append((c - 10, c + 10, 0.06 + rng.random() * 0.12))
    target_zones = [
        (length_m * 0.62, length_m * 0.7, "victim_cluster_a"),
        (length_m * 0.78, length_m * 0.9, "victim_cluster_b"),
    ]
    return TunnelGeometry(
        length_m=length_m,
        width_m=width_m,
        height_m=5.0,
        entrance_s=0.0,
        segments=segments,
        collapse_points=collapse_points,
        relay_anchor_zones=relay_anchor_zones,
        signal_shadow_zones=signal_shadow_zones,
        target_zones=target_zones,
    )


def segment_loss_mul(geom: TunnelGeometry, s: float) -> float:
    m = 1.0
    for seg in geom.segments:
        if seg.start_s <= s <= seg.end_s:
            m *= seg.attenuation_mul
            if seg.blocked:
                m *= 1.28
    for z0, z1, loss_add in geom.signal_shadow_zones:
        if z0 <= s <= z1:
            m *= 1.0 + loss_add
    return m
